src/refdb_reftable.c

```diff
@@ -0,0 +1,513 @@
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "git2/sys/refdb_backend.h"
+#include "git2/sys/refs.h"
+#include "reflog.h"
+#include "signature.h"
+#include "wildmatch.h"
+#include "../reftable/reftable.h"
+
+typedef struct {
+	git_refdb_backend parent;
+	git_repository *repo;
+	struct stack *stack;
+} reftable_backend;
+
+typedef struct {
+	git_reference_iterator parent;
+	struct merged_table *table;
+	struct iterator iter;
+	const char *glob;
+} reftable_iterator;
+
+typedef struct {
+	struct ref_record *refs;
+	struct log_record *logs;
+	size_t nrefs;
+	size_t nlogs;
+	uint64_t index;
+} reftable_writer_arg;
+
+static int reftable_reference_from_record(git_reference **out, struct ref_record *record)
+{
+	git_reference *ref;
+	git_oid oid, peeled;
+	int error;
+
+	if (record->target) {
+		ref = git_reference__alloc_symbolic(record->ref_name, record->target);
+	} else {
+		if ((error = git_oid_fromraw(&oid, record->value)) < 0 ||
+		    (error = git_oid_fromraw(&oid, record->target_value)) < 0)
+			goto out;
+		ref = git_reference__alloc(record->ref_name, &oid, &peeled);
+	}
+	GIT_ERROR_CHECK_ALLOC(ref);
+
+	*out = ref;
+out:
+	return error;
+}
+
+static int reftable_log_fill(struct log_record *out,
+			     const git_signature *who,
+			     const git_oid *old_id,
+			     const git_oid *new_id,
+			     const char *reference,
+			     const char *message)
+{
+	memset(out, 0, sizeof(*out));
+	if (who) {
+		out->name = who->name;
+		out->email = who->email;
+		out->time = who->when.time;
+		out->tz_offset = who->when.offset;
+	}
+	if (old_id)
+		out->old_hash = (unsigned char *) old_id->id;
+	if (new_id)
+		out->new_hash = (unsigned char *) new_id->id;
+	out->ref_name = (char *) reference;
+	out->message = (char *) message;
+	return 0;
+}
+
+static int reftable_writer(struct writer *wr, void *_arg)
+{
+	reftable_writer_arg *arg = (reftable_writer_arg *) _arg;
+	writer_set_limits(wr, arg->index, arg->index + arg->nrefs + arg->nlogs);
+	if (writer_add_refs(wr, arg->refs, arg->nrefs) < 0)
+		return -1;
+	if (writer_add_logs(wr, arg->logs, arg->nlogs) < 0)
+		return -1;
+	return 0;
+}
+
+static int reftable_write_records(git_refdb_backend *_backend,
+				  struct ref_record *refs, int nrefs,
+				  struct log_record *logs, int nlogs)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	reftable_writer_arg arg = {
+		refs, logs, nrefs, nlogs, stack_next_update_index(backend->stack)
+	};
+
+	if (stack_add(backend->stack, reftable_writer, &arg) < 0)
+		return -1;
+
+	return 0;
+}
+
+static int reftable_exists(
+	int *exists,
+	git_refdb_backend *_backend,
+	const char *ref_name)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record ref = { 0 };
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, ref_name, &ref)) < 0)
```

Comment:
I'd probably have to call `stack_reload` everytime we call into the refdb as we don't know whether any concurrent process has modified the reftable and as libgit2 may be a long-running process. Is this going to be a performance hog or do you expect it to work just fine?

Reply:
It will read the reftable list, which is a small file, so it is relatively cheap. 

However, the reftable instance you hold will give you a consistent snapshot of the refdb at some point in time, so it's up to you to decide whether you really need this. If there is a concurrent process, you'd normally have some sort of synchronization with that process, no?

Reply:
> If there is a concurrent process, you'd normally have some sort of synchronization with that process, no?

Unfortunately, no. As explained below, libgit2 may be part of a long running application like e.g. an IDE where you don't know what other tools the user may be running. So we have to consistently refresh all files that could potentially change, like references or the config file.